components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/JDBCPermissionBasedInternalScopeValidator.java

```diff
@@ -167,14 +167,14 @@ private List<Scope> getUserAllowedScopes(AuthenticatedUser authenticatedUser, St
             int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
 
             Set<Scope> allScopes = getScopesOfPermissionType(tenantId);
-            List<Scope> allScopesList = new ArrayList<>(allScopes);
-
             if (ArrayUtils.contains(requestedScopes, SYSTEM_SCOPE)) {
+                List<Scope> allScopesList = new ArrayList<>(allScopes);
                 requestedScopes = getScopeNames(allScopesList);
             } else {
                 // filter out the internal scopes
                 requestedScopes = Oauth2ScopeUtils.getRequestedScopes(requestedScopes);
             }
+            Set<String> requestedScopesSet = new HashSet<>(Arrays.asList(requestedScopes));
```

Comment:
Can't we build the set directly in these two methods?
We might have to modify getScopeNames() and Oauth2ScopeUtils.getRequestedScopes()

Check if its worth the effort over copying once. Otherwise this is fine.

Also how do we handle the case "SYSTEM some_random_scope"
some_random_scope is not an internal_* type scope but another scope which is registered and allowed for the app. Do they type of scopes get filtered before moving to this logic?

Reply:
These functions are only for handling internal scopes. other registered scopes are handled in the parent function.